docs/config.md
Configuration File
==================
Below is a detailed description of what a value does in the configuration file.

!!! Important
    A tick is 1/20 of a second.
    A bucket holds 1000 millibuckets

## barrelNumberOfBuckets
The number of buckets of fluid a barrel will hold.
*Must be an integer greater than 0.*

## rainFillAmount
The number of millibuckets per tick that a barrel will fill with water per tick.
*Must be an integer greater than 0.*

## woodBarrelMaxTemp
The maximum temperature of a fluid the wooden can accept. Water is 300 and lava is 1300. Anything above 433 is considered hot.
*Must be an integer greater than or equal to 0.*


## secondsToSpawnMobs
The number of seconds that it takes to spawn mobs in a barrel.
*Must be an integer greater than 0.*

## secondsToCompost
The number of seconds that it takes to compost solids in a barrel once its full.
*Must be an integer greater than 0.*

## maxSolidAmount
The max solid amount that a barrel will hold. When the solid amount in a barrel reaches this number, it will start composting. This solid amount will be the equivalent of one dirt block.
*Must be an integer greater than 0.*

## secondsToTransformFluid
The number of seconds that it takes to transform a fluid in a barrel.
*Must be an integer greater than 0.*

## maxBonusStringCount
The maximum number of bonus string that an infested leaf will drop when broken with a crook.
*Must be an integer greater than 0.*

## minStringCount
The maximum number of bonus string that an infested leaf will drop when broken with a crook.
*Must be an integer greater than or equal to 0.*

## vanillaDropSimulateCount
The number of times a crook will simulate breaking a block to get its vanilla drops.
*Must be an integer greater than 0.*

## ticksBetweenMelts
The crucible does not perform a melt operation every tick. This config allows the user to define how many ticks take place in between operations. If it is set to 1, it will perform an operation every tick.
*Must be an integer greater than 0.*

## crucibleNumberOfBuckets
The number of buckets that the crucible will hold.
*Must be a number greater than 0.*

## woodHeatRate
The Wooden Crucible does not use the Heat Registry. Regardless of the heat source placed below it, the Wooden Crucible will melt items down at the rate of this config.
*Must be a number greater than 0.*

## secondsToTransformLeaves
The number of seconds that it takes to completely infest leaves.
*Must be an integer greater than 0.*

## spreadChance
The percent chance that an infested leaf block will spread to a non-infested leaf block next to it.
*Must be a value between 0.001 and 1.0.*

## ticksBetweenSpreadAttempt
Infested leaves do not attempt to spread every tick. This config allows the user to define how many ticks take place in between operations. If it is set to 1, infested leaves will attempt to spread every tick.
*Must be a value greater than 0.*

## flattenSieveRecipes
Sets whether sieves will get the results for all the meshes below the one currently in the sieve. For example, if an iron mesh is in the sieve, it will get results for the iron, flint and string meshes.
*Must be `true` or `false`*.

## sieveRadius
The number of blocks away that the sieve will try to find other sieves to activate. For example, if set to 2, the sieve will look in a 5 x 5 block area (with the sieve that was activated in the center) for other sieves. *Must be a value between 0 and 5*.

## enableMeshDurability
This enables meshes to be breakable. *Must be `true` or `false`*.

## enableThermalExpansion
Enables ore drops for Thermal Expansion ores

## enableImmersiveEngineering
Enables ore drops for Immersive Engineering ores

## enableMekanism
Enables ore drops for Mekanism ores

## enableCreate
Enables ore drops for Create ores

## enableSilentMechanism
Enables ore drops for Silent Mechanism ores

## enableDebugLogging
Enables extra logging.

src/main/java/novamachina/exnihilosequentia/common/utility/Config.java
package novamachina.exnihilosequentia.common.utility;

import com.electronwill.nightconfig.core.file.CommentedFileConfig;
import com.electronwill.nightconfig.core.io.WritingMode;
import net.minecraftforge.common.ForgeConfigSpec;
import net.minecraftforge.common.ForgeConfigSpec.Builder;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.config.ModConfig;

import java.nio.file.Path;

@Mod.EventBusSubscriber
public class Config {

    private static final String CATEGORY_BARREL = "barrel";
    private static final String CATEGORY_CROOK = "crook";
    private static final String CATEGORY_CRUCIBLE = "crucible";
    private static final String CATEGORY_INFESTED_LEAVES = "infested_leaves";
    private static final String CATEGORY_SIEVE = "sieve";
    private static final String CATEGORY_COMPAT = "compatibility";
    private static final String CATEGORY_DEBUG = "debug";
    private static final String SUBCATEGORY_BARREL_MOB = "mob_spawn";
    private static final String SUBCATEGORY_BARREL_COMPOST = "compost";
    private static final String SUBCATEGORY_BARREL_FLUID = "fluid_transform";
    private static final String SUBCATEGORY_CRUCIBLE_WOOD = "wood";

    private static final ForgeConfigSpec.Builder COMMON_BUILDER = new Builder();

    public static ForgeConfigSpec COMMON_CONFIG;

    // Barrel
    public static ForgeConfigSpec.IntValue SECONDS_TO_SPAWN;
    public static ForgeConfigSpec.IntValue BARREL_MAX_SOLID_AMOUNT;
    public static ForgeConfigSpec.IntValue SECONDS_TO_COMPOST;
    public static ForgeConfigSpec.IntValue SECONDS_TO_FLUID_TRANSFORM;
    public static ForgeConfigSpec.IntValue BARREL_NUMBER_OF_BUCKETS;
    public static ForgeConfigSpec.IntValue RAIN_FILL_AMOUNT;
    public static ForgeConfigSpec.IntValue WOOD_BARREL_MAX_TEMP;

    // Crook
    public static ForgeConfigSpec.IntValue MAX_BONUS_STRING_COUNT;
    public static ForgeConfigSpec.IntValue MIN_STRING_COUNT;
    public static ForgeConfigSpec.IntValue VANILLA_SIMULATE_DROP_COUNT;

    // Crucible
    public static ForgeConfigSpec.IntValue TICKS_BETWEEN_MELTS;
    public static ForgeConfigSpec.IntValue CRUCIBLE_NUMBER_OF_BUCKETS;
    public static ForgeConfigSpec.IntValue WOOD_HEAT_RATE;

    // Infested Leaves
    public static ForgeConfigSpec.IntValue SECONDS_TO_TRANSFORM_LEAVES;
    public static ForgeConfigSpec.DoubleValue SPREAD_CHANCE;
    public static ForgeConfigSpec.IntValue TICKS_BETWEEN_SPREAD_ATTEMPT;

    // Sieve
    public static ForgeConfigSpec.BooleanValue FLATTEN_SIEVE_RECIPES;
    public static ForgeConfigSpec.IntValue SIEVE_RANGE;
    public static ForgeConfigSpec.BooleanValue ENABLE_MESH_DURABILITY;

    // Compat
    public static ForgeConfigSpec.BooleanValue ENABLE_THERMAL;
    public static ForgeConfigSpec.BooleanValue ENABLE_IMMERSIVE;
    public static ForgeConfigSpec.BooleanValue ENABLE_MEKANISM;
    public static ForgeConfigSpec.BooleanValue ENABLE_CREATE;
    public static ForgeConfigSpec.BooleanValue ENABLE_SILENT;

    // Debugging
    public static ForgeConfigSpec.BooleanValue ENABLE_DEBUG_LOGGING;

    static {
        COMMON_BUILDER.comment("Barrel Configs").push(CATEGORY_BARREL);
        barrelConfigs();
        COMMON_BUILDER.pop();
        COMMON_BUILDER.comment("Crook Configs").push(CATEGORY_CROOK);
        crookConfigs();
        COMMON_BUILDER.pop();
        COMMON_BUILDER.comment("Crucible Configs").push(CATEGORY_CRUCIBLE);
        crucibleConfigs();
        COMMON_BUILDER.pop();
        COMMON_BUILDER.comment("Infested Leaves Configs").push(CATEGORY_INFESTED_LEAVES);
        infestedLeavesConfigs();
        COMMON_BUILDER.pop();
        COMMON_BUILDER.comment("Sieve Configs").push(CATEGORY_SIEVE);
        sieveConfigs();
        COMMON_BUILDER.pop();
        COMMON_BUILDER.comment("Compatibility Configs").push(CATEGORY_COMPAT);
        compatConfigs();
        COMMON_BUILDER.pop();
        COMMON_BUILDER.comment("Debug Configs").push(CATEGORY_DEBUG);
        debugConfigs();
        COMMON_BUILDER.pop();

        COMMON_CONFIG = COMMON_BUILDER.build();
    }

    private static void debugConfigs() {
        ENABLE_DEBUG_LOGGING = COMMON_BUILDER.comment("Enable extra logging? (Default: false)")
            .define("enableDebugLogging", false);
    }

    private static void compatConfigs() {
        ENABLE_THERMAL = COMMON_BUILDER.comment("Should Thermal Expansion ores be enabled? (Default: false)")
            .define("enableThermalExpansion", false);
        ENABLE_IMMERSIVE = COMMON_BUILDER.comment("Should Immersive Engineering ores be enabled? (Default: false)")
            .define("enableImmersiveEngineering", false);
        ENABLE_MEKANISM = COMMON_BUILDER.comment("Should Mekanism ores be enabled? (Default: false)")
            .define("enableMekanism", false);
        ENABLE_CREATE = COMMON_BUILDER.comment("Should Create ores be enabled? (Default: false)")
            .define("enableCreate", false);
        ENABLE_SILENT = COMMON_BUILDER.comment("Should Silent Mechanism ores be enabled? (Default: false)")
            .define("enableSilentMechanism", false);
    }

    private static void sieveConfigs() {
        FLATTEN_SIEVE_RECIPES = COMMON_BUILDER
            .comment("Sieve will get results for all mesh tiers below the one in the sieve (Default: true)")
            .define("flattenSieveRecipes", true);
        SIEVE_RANGE = COMMON_BUILDER
            .comment("Defines the radius that a sieve will attempt to activate other sieves (Default: 2)")
            .defineInRange("sieveRange", 2, 0, 5);
        ENABLE_MESH_DURABILITY = COMMON_BUILDER
            .comment("Meshes will have durability and can break. (Default: false)")
            .define("enableMeshDurability", false);
    }

    private static void infestedLeavesConfigs() {
        SECONDS_TO_TRANSFORM_LEAVES = COMMON_BUILDER
            .comment("Number of seconds to for leaves to become completely infested (Default: 10)")
            .defineInRange("secondsToTransformLeaves", 10, 1, Integer.MAX_VALUE);
        SPREAD_CHANCE = COMMON_BUILDER.comment("Percentage of the time that infested leaves will spread (Default: 0.3)")
            .defineInRange("spreadChance", 0.3, 0.001, 1.0);
        TICKS_BETWEEN_SPREAD_ATTEMPT = COMMON_BUILDER
            .comment("Number of ticks between infested leave spread attempts (Default: 100)")
            .defineInRange("ticksBetweenSpreadAttempt", 100, 1, Integer.MAX_VALUE);
    }

    private static void crucibleConfigs() {
        TICKS_BETWEEN_MELTS = COMMON_BUILDER.comment("Ticks between melting operations (Default: 20)")
            .defineInRange("ticksBetweenMelts", 20, 1, Integer.MAX_VALUE);
        CRUCIBLE_NUMBER_OF_BUCKETS = COMMON_BUILDER.comment("Number of buckets the crucible will hold (Default: 4)")
            .defineInRange("crucibleNumberOfBuckets", 4, 1, Integer.MAX_VALUE);

        COMMON_BUILDER.comment("Wooden Crucible Configs").push(SUBCATEGORY_CRUCIBLE_WOOD);
        WOOD_HEAT_RATE = COMMON_BUILDER
            .comment("Heat rate the Wood Crucible will use regardless of heat source below (Default: 2)")
            .defineInRange("woodHeatRate", 2, 1, Integer.MAX_VALUE);
        COMMON_BUILDER.pop();
    }

    private static void crookConfigs() {
        MAX_BONUS_STRING_COUNT = COMMON_BUILDER
            .comment("Maximum additional string that a crook will drop from infested leaves in addition to the minimum string count (Default: 3)")
            .defineInRange("maxBonusStringCount", 3, 0, Integer.MAX_VALUE);
        MIN_STRING_COUNT = COMMON_BUILDER
            .comment("Minimum string that a crook will drop from infested leaves (Default: 2)")
            .defineInRange("minStringCount", 2, 1, Integer.MAX_VALUE);
        VANILLA_SIMULATE_DROP_COUNT = COMMON_BUILDER
            .comment("Number of times the crook will \"break\" a leaf block to get drops (Default: 3)")
            .defineInRange("vanillaDropSimulateCount", 3, 1, Integer.MAX_VALUE);
    }

    private static void barrelConfigs() {
        BARREL_NUMBER_OF_BUCKETS = COMMON_BUILDER.comment("Number of buckets the barrel will hold (Default: 1)")
            .defineInRange("barrelNumberOfBuckets", 1, 1, Integer.MAX_VALUE);
        RAIN_FILL_AMOUNT = COMMON_BUILDER.comment("How much fluid rain will fill per iteration (Default: 2)")
            .defineInRange("rainFillAmount", 2, 1, Integer.MAX_VALUE);
        WOOD_BARREL_MAX_TEMP = COMMON_BUILDER.comment("The max temperature a barrel can accept; water is 300 (Default: 300)")
            .defineInRange("woodBarrelMaxTemp", 300, 0, Integer.MAX_VALUE);

        COMMON_BUILDER.comment("Mob Spawn Configs").push(SUBCATEGORY_BARREL_MOB);
        SECONDS_TO_SPAWN = COMMON_BUILDER.comment("Number of seconds to spawn mobs (Default: 10)")
            .defineInRange("secondsToSpawnMobs", 10, 1, Integer.MAX_VALUE);
        COMMON_BUILDER.pop();

        COMMON_BUILDER.comment("Compost Configs").push(SUBCATEGORY_BARREL_COMPOST);
        BARREL_MAX_SOLID_AMOUNT = COMMON_BUILDER
            .comment("How much solids need to be in barrel before composting starts (Default: 1000)")
            .defineInRange("maxSolidAmount", 1000, 1, Integer.MAX_VALUE);
        SECONDS_TO_COMPOST = COMMON_BUILDER.comment("Number of seconds to spawn mobs (Default: 10)")
            .defineInRange("secondsToCompost", 10, 1, Integer.MAX_VALUE);
        COMMON_BUILDER.pop();

        COMMON_BUILDER.comment("Fluid Transform Configs").push(SUBCATEGORY_BARREL_FLUID);
        SECONDS_TO_FLUID_TRANSFORM = COMMON_BUILDER.comment("Number of seconds to transform fluids (Default: 10)")
            .defineInRange("secondsToTransformFluid", 10, 1, Integer.MAX_VALUE);
        COMMON_BUILDER.pop();
    }

    public static void loadConfig(ForgeConfigSpec spec, Path path) {
        final CommentedFileConfig configData = CommentedFileConfig.builder(path).sync().autosave()
            .writingMode(
                WritingMode.REPLACE).build();

        configData.load();
        spec.setConfig(configData);
    }

    @SubscribeEvent
    public static void onLoad(final ModConfig.Loading configEvent) {
    }

    @SubscribeEvent
    public static void onReload(final ModConfig.Reloading configEvent) {
    }
}
